### ISSUE TITLE
Please add "dim" feature for function "torch.masked_select"

### ISSUE THREAD

**Issue by songyuc** [pytorch/pytorch#48830]
## 🚀 Feature

A `dim` feature can be added for function `torch.masked_select`.

## Motivation

I think it will be more convenient if we can have a "dim" parameter in `torch.masked_select`.
There is a scenario as we have these three tensors,
labels: [N]
scores: [N]
boxes: [N,4],
and we can get the indexes like, `bool_mask = labels== 2`.
Then we can select `labels` and `scores` as,
```python
labels= labels.masked_select(bool_mask)
scores= scores.masked_select(bool_mask)
```
So, I want to also select the boxes as,
```python
boxes= boxes.masked_select(bool_mask)
# boxes: tensor, [M,4]
```

**Comment by ailzhang** [comment]
Not really directly related to numpy, but I feel this falls into a similar API design, please feel free to move the label if it's not accurate. Thanks!

**Comment by mruberry** [comment]
Thank you for the suggestion, @songyuc. What would be the simplest way to do this currently? Would it be by taking a view of the tensor before performed the masked selection?

**Comment by songyuc** [comment]
Hi, @mruberry, I just do it like,
```python
boxes= boxes[bool_mask,:]
# boxes: tensor, [M,4]
```
which requires me to fill all the other dimensions.

**Comment by songyuc** [comment]
@mruberry , by the way, what is the difference between `tensor.masked_select(mask)` and `tensor[mask]`

**Comment by mruberry** [comment]
> @mruberry , by the way, what is the difference between `tensor.masked_select(mask)` and `tensor[mask]`

Good question! I don't know offhand, sorry. We'd have to dig through the code.

**Comment by rpmcruz** [comment]
> @mruberry , by the way, what is the difference between `tensor.masked_select(mask)` and `tensor[mask]`

I think the only difference is that `tensor.masked_select` is broadcastable. That is, if you have a tensor that is (A,B,C,D) and a mask that is (A,1,1,D) then `tensor.masked_select` will work, but `tensor[mask]` will not. This is quite useful in some neural networks losses.

**Comment by davzaman** [comment]
I am not sure what the simplest way to do this would be but I am actively interested in this feature.  Currently numpy supports this with something like: 
```python
masked_array(
  data=[[--, 1.7950215339660645, -0.007744395174086094,
         -0.42541494965553284],
        [2.5049524307250977, 0.013974136672914028, -0.9842795133590698,
         -0.25598782300949097],
        [-0.7021965980529785, -0.551115095615387, 0.9622372984886169,
         0.07153400033712387]],
  mask=[[ True, False, False, False],
        [False, False, False, False],
        [False, False, False, False]],
  fill_value=1e+20,
  dtype=float32)
```
The importance here is that it allows me to mask unevenly and apply functions to that. So If I want to compute the mean first in one dimension and then the other, then it can do that. As opposed to an element-wise mean as `masked_select` would do if I used that (since it collapses its selection to 1D). 

The use case for me is assessing metrics only for certain values. For example I have missing values in certain spots, and I want to evaluate the output only where the data were originally missing, which is absolutely not uniform across features and samples. Right now I am using numpy which is expensive and slow because I have to transfer all my data to the cpu to do the computation for billions of rows of data. I cannot use `masked_select` because I do not want an element-wise metric, I want a column-wise one.

**Comment by BarunKGP** [comment]
I'd also be interested in a feature like this. Are there any plans to add this functionality in future versions?
